source/distributor/service_impl.cc

```diff
@@ -0,0 +1,119 @@
+#include "distributor/service_impl.h"
+
+#include <grpc++/grpc++.h>
+
+#include "envoy/config/core/v3/base.pb.h"
+
+#include "external/envoy/source/common/common/assert.h"
+#include "external/envoy/source/common/protobuf/message_validator_impl.h"
+#include "external/envoy/source/common/protobuf/utility.h"
+
+#include "api/distributor/distributor.pb.validate.h"
+
+namespace Nighthawk {
+namespace {
+
+grpc::Status validateRequest(const nighthawk::DistributedRequest& request) {
+  Envoy::ProtobufMessage::ValidationVisitor& validation_visitor =
+      Envoy::ProtobufMessage::getStrictValidationVisitor();
+  try {
+    Envoy::MessageUtil::validate(request, validation_visitor);
+  } catch (const Envoy::ProtoValidationException& e) {
+    return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, e.what());
+  }
+
+  if (request.has_execution_request()) {
+    const nighthawk::client::ExecutionRequest& execution_request = request.execution_request();
+    if (!execution_request.start_request().has_options()) {
+      return grpc::Status(
+          grpc::StatusCode::INVALID_ARGUMENT,
+          "DistributedRequest.ExecutionRequest.StartRequest MUST have CommandLineOptions.");
+    }
+  } else {
+    return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
+                        "DistributedRequest.ExecutionRequest MUST be specified.");
+  }
+  return grpc::Status::OK;
+}
+
+} // namespace
+
+absl::StatusOr<nighthawk::client::ExecutionResponse>
+NighthawkDistributorServiceImpl::handleExecutionRequest(
+    const envoy::config::core::v3::Address& service,
+    const nighthawk::client::ExecutionRequest& request) const {
+  RELEASE_ASSERT(service_client_ != nullptr, "service_client_ != nullptr");
+  std::shared_ptr<grpc::Channel> channel;
+  channel = grpc::CreateChannel(fmt::format("{}:{}", service.socket_address().address(),
+                                            service.socket_address().port_value()),
+                                grpc::InsecureChannelCredentials());
+  std::unique_ptr<nighthawk::client::NighthawkService::Stub> stub =
+      std::make_unique<nighthawk::client::NighthawkService::Stub>(channel);
+  return service_client_->PerformNighthawkBenchmark(stub.get(), request.start_request().options());
+}
+
+// Translates one or more backend response into a single reply message
+std::tuple<grpc::Status, nighthawk::DistributedResponse>
+NighthawkDistributorServiceImpl::handleRequest(const nighthawk::DistributedRequest& request) const {
+  RELEASE_ASSERT(request.has_execution_request(), "request.has_execution_request()");
```

Comment:
Why do we prefer to have RELEASE_ASSERTs as compared to validating the request and returning an error?

Assuming the asserts crash the code, it may be ok to do that when verifying internal correctness, but we probably don't want to do that for an invalid argument.

Reply:
This is pre-checked in `validateRequest()`, hence this assert triggering is indicative of something being very wrong.
The flow is covered in a test: https://github.com/envoyproxy/nighthawk/pull/669/files/5462844f69719621185528e77cb583c66f21900a#diff-1b41e54eeb708a749bd8c54a830c84409cecdd778b7e578dbcd45dc575e1416fR87
Given that context, does that address your concern?

Reply:
Following our offline discussion, removed the asserts is d0fb1b4c. Thanks for persisting @mum4k, this is overly cautious, and not sustainable as a strategy. Thanks for pointing that out.

Reply:
For posterity - discussed offline, the asserts are considered to be too defensive, the unit tests in place already verify the validation occurs before the code handles the request.

Otto please indicate if I misinterpreted anything.